### ISSUE TITLE
Where with parametrized INTERVAL

### ISSUE THREAD

**Issue by ikawalec** [Masterminds/squirrel#379]
Hi,

I'm having an issue with using where expr with parametrized INTERVAL.

Code:

```
d := 3

builder := r.StatementBuilderType.
  Delete("test").
  Where(squirrel.Expr("expiry <= now()::timestamptz - INTERVAL '? DAYS'", d))
		
query, args, err = builder.ToSql()
	
logrus.Infof("query: %s, args: %+v", query, args)
	
// db execute
```

I'm getting the following error:

```
ERROR: could not parse "$1 DAYS" as type interval: interval: missing number at position 0: "$1 DAYS" (SQLSTATE 22007)
```

Generated query and args:

```
query: DELETE FROM test WHERE expiry \u003c= now()::timestamptz - INTERVAL '$1 DAYS'
args: [3]
```

Expected SQL:

```
delete from test where expiry <= now()::timestamptz - INTERVAL '3 days';
```

Is this a bug? Is there any other way this can be achieved?

Thanks!

**Comment by lann** [comment]
Assuming you are using postgres, I believe the quotes are the problem: `INTERVAL '? DAYS'` -> `INTERVAL ? DAYS`

**Comment by ikawalec** [comment]
> Assuming you are using postgres, I believe the quotes are the problem: `INTERVAL '? DAYS'` -> `INTERVAL ? DAYS`

That's correct, I'm using postgres. Without quotes, it's not a valid SQL statement:

```
delete from test where expiry <= now()::timestamptz - INTERVAL 3 days;

at or near "3": syntax error
DETAIL:  source SQL:
delete from test where expiry <= now()::timestamptz - INTERVAL 3 days
```

**Comment by lann** [comment]
It looks like postgres expects the number to be a string, so you may also need to e.g. `strconv.Itoa` the Go int before passing it to the database.

**Comment by ikawalec** [comment]
Thanks but still even with days converted using`strconv.Itoa` I still get the same error (assuming quotes are mandatory).

```
squirrel.Expr("expiry <= now()::timestamptz - INTERVAL '? DAYS'", strconv.Itoa(days)),
```

```
ERROR: could not parse "$1 DAYS" as type interval: interval: missing number at position 0: "$1 DAYS" (SQLSTATE 22007)
```

**Comment by lann** [comment]
The quotes mean you are passing the literal string "? DAYS"; Squirrel converts the `?` to `$1` because its placeholder replacement doesn't know how to deal with strings.

I believe what you need is:
```go
squirrel.Expr("expiry <= now()::timestamptz - INTERVAL ? DAYS", strconv.Itoa(days)),

**Comment by ikawalec** [comment]
> The quotes mean you are passing the literal string "? DAYS"; Squirrel converts the `?` to `$1` because its placeholder replacement doesn't know how to deal with strings.
> 
> I believe what you need is:
> 
> ```go
> squirrel.Expr("expiry <= now()::timestamptz - INTERVAL ? DAYS", strconv.Itoa(days)),
> ```

The example that you provided will generate the following SQL query:

```
DELETE FROM test WHERE expiry \u003c= now()::timestamptz - INTERVAL 3 DAYS
```

In order for that to work with postgres: https://www.postgresql.org/docs/8.0/functions-datetime.html, the INTERVAL should use quotes `'3 DAYS'`:

```
DELETE FROM test WHERE expiry \u003c= now()::timestamptz - INTERVAL '3 DAYS'
```

**Comment by lann** [comment]
Ah sorry, I got bad information from Google. In that case I would expect something like this to work:

```go
squirrel.Expr("expiry <= now()::timestamptz - INTERVAL ?", fmt.Sprintf("%d DAYS", days))

**Comment by ikawalec** [comment]
> Ah sorry, I got bad information from Google. In that case I would expect something like this to work:
> 
> ```go
> squirrel.Expr("expiry <= now()::timestamptz - INTERVAL ?", fmt.Sprintf("%d DAYS", days))
> ```

Unfortunately, this would also fail:

```
 query: DELETE FROM test WHERE expiry \u003c= now()::timestamptz - INTERVAL $1
 args: [3 DAYS]
 error: ERROR: at or near "1": syntax error (SQLSTATE 42601)
 ```

**Comment by efitzkiwi** [comment]
Only way I've gotten this to work is by doing something such as the following
```
Where(`"timestamp" >= (recent_close."timestamp" - interval '1 day' * ?)`, finalInterval)
```
Where finalInterval is a number.

The key part being the multiplier, unfortunately seems like a driver issue.
